Bran Ladder-step3

june20516 opened the pull request:
요구 사항을 반영하려 노력하며 작성했습니다.

- 기능별 구조체 `InputView` , `LadderGame` , `PrintView`

- 저장 구조체 `LadderPlayer` , `LadderStep`

- 사다리 위에 이름을 출력

june20516 committed:
구조체로 나누어서 사다리게임 재구현

- LadderPlayer : 참가자 저장 struct
- LadderStep : 사다리 가로요소 한개. 랜덤bool값을 갖도록 초기화
- InputView : 값을 입력받아 사용 가능하게 변환하는 struct
- LadderGame : InputView에서 넘겨받은 값으로 데이터를 만드는 logic Struct
- PrintView : 데이터에 따라 게임요소를 출력하는 struct

june20516 committed:
step 파일 디렉터리 정리

june20516 committed:
주석 정리 및 스코어 확인

june20516 committed:
LadderStep 중복 요소 방지 구현

초기 Step을 만들어두고,
이전 step을 검사하여 겹치지않는 step을 만드는 checkPriorStep 함수 생성

june20516 committed:
LadderPlayers 이름 출력

글자수별 마진값을 배열로 보관, 이름 사이즈에 따라 마진 출력하여 이름 정렬.

june20516 committed:
실행 함수 작성

june20516 committed:
파일 및 디렉터리 정리

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,54 @@
+    
+    //프로퍼티 선언
+    let players = getPlayers()
+    let height = getHeight()
```
let 으로 하더라도 struct 내부는 private 우선적으로 고려하세요.

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -0,0 +1,54 @@
+            players.append(LadderPlayer(name: stringName))
+        }
+        print(players.count)
+        return players
```
static 으로 선언한 타입 메소드들인데 마치 인스턴스 변수 players에 접근하는 것처럼 오해를 주세요.
static 으로 만들었으면 위에 players, height 가 필요없는건지, 
필요한 값이라면 static 함수를 이용하지 않을지 구분해보세요.
지금 let 도 있고 static 으로 선언한 것은 나쁜 구조네요. 
여기 있는 print()도 삭제하세요.

june20516 replied:
피드백 반영하여 코드 수정해보았습니다.


**InputView**
- static을 사용하지 않고, 변수를 없애고 method의 반환값으로만 입력값을 가져올 수 있게 하였습니다.

**LadderStep**
- 변수명을 의미가 전달 될 수 있도록 수정했습니다.
- 변수의 값을 바꿀 수 있도록 메서드를 개선했습니다.

**PrintView**
- step 그리기 요소의 enum이 불필요하다고 판단하였기 때문에, enum을 빼고 dictionary 하나만 가지는 구조로 수정하였습니다. 
- 

**LadderGameError** / **main**
- 에러및 예외를 처리하기 위한 부분을 추가했습니다. 
- 예측 가능한 에러가 많지 않아, throw catch 를 사용하지 않고 string enum으로 구성하였습니다.

godrm replied:
전체 내용은 개별 댓글이 아니라 전체 댓글로 남겨주세요 ㅎㅎ

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,47 @@
+
+struct LadderGame {
+    // 입력값을 가져옴
+    let inputValue = InputView()
```
LadderGame 과 InputView (OutputView나 ResultView도 마찬가지로) 관계를 항상 분리하세요. 
입력, 출력을 담당하는 객체를 이렇게 내부에 포함하는 구조는 지양하세요.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,47 @@
+        var oneLine = [LadderStep()]
+        //중복 안되게 생성하여 배열에 추가
+        for stepIndex in 0..<inputValue.players.count-2 {
+            oneLine.append(checkPriorStep(lastStep: oneLine[stepIndex]))
```
함수를 중첩해서 호출하는 것도 코드 흐름상 좋지 않으니까 풀어서 쓰세요.
한 줄로 쓴다고 해서 좋은게 아닙니다.

godrm commented on LadderGame/LadderGame/LadderPlayer.swift:
```diff
@@ -0,0 +1,17 @@
+    let name: String
+    init(name: String) {
+        self.name = name
+    }
```
지금처럼 이렇게 로직이 없는 경우는 let 정도로 선언해도 상관없습니다. 참고하세요.

godrm commented on LadderGame/LadderGame/LadderStep.swift:
```diff
@@ -0,0 +1,24 @@
+//사다리 한개
+struct LadderStep {
+    //사다리 요소
+    var dataStep: Bool
```
dataStep 은 무엇을 위한 변수인가요? 이름으로 유추하기 어렵네요. 
data- 라는 단어도 `Data` 라는 타입이 있어서 붙이면 오해하기 쉬운 단어입니다. 차라리 빼세요.
private 을 우선적으로 고려하세요.

godrm commented on LadderGame/LadderGame/PrintView.swift:
```diff
@@ -0,0 +1,57 @@
+        for step in oneFloor {
+            // enum 언래핑
+            guard let haveOrNone = drawElements[step.dataStep] else {
+                return "error"
```
특정한 상황이나 에러값을 지금처럼 `"error"` 형태로 특정한 값으로 넘기는 것은 좋지 않습니다.
차라리 String? 옵셔널로 처리하거나 exception 을 활용하는 게 낫습니다.

june20516 committed:
디렉토리 정리

june20516 committed:
InputView.swift 접근범위 재설정

프로퍼티를 삭제하고 필요한 값을 인터널 메서드로 꺼내도록 변경

june20516 committed:
LadderStep 구조체 인스턴스 범위 변경

june20516 committed:
ladderStep 처리 방식 변경

- 열거형 rawvalue 참조방식에서 dic 단일 구조로 변경
- 옵셔널 처리 방식 변경 : Error 객체 생성, 테스트 코드 작성 후 errorhadling 적용 예정

june20516 committed:
ladderstep 메서드 변경

stepDecision을 변경하는 mutating func로 변경.

june20516 committed:
Error 및 예외처리

입력받은 값을 체크하여 잘못된 경우 프로그램 중단

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,51 @@
+            return nextStep
+        }
+        var nextStep = LadderStep()
+        nextStep.makeRandomStep()
```
LadderStep 에 static 하게 emptyStep과 randomeStep을 만들 수 있으면 하나로 가능하지 않을까요?

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,145 +8,25 @@
+    let output = PrintView()
+    let completedLadder = game.completeWholeLadder()
+    output.orderPlayerNames(players: game.players)
+    output.printWholeLine(wholeLadder: completedLadder)
```
어떤 메소드는 orderPlayerNames() 이고 어떤 메소도는 printWholeLine() 인데
비슷한 역할을 하는 것 같은데, order- 와 print- 를 나눈 기준은 뭔가요?

june20516 replied:
공백을 넣어 보기좋게 정렬한다는 의미로 나누었었으나, 프린트의 과정일 뿐이라 생각하여 print로 수정하여 반영했습니다.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -0,0 +1,51 @@
+struct LadderGame {
+    // 입력값을 가져옴
+    let players: [LadderPlayer]
+    let height: Int
```
let 으로 하더라도 private 지정하는 방식을 우선적으로 고려하세요.

june20516 committed:
객체 구조 변경 및 LadderError protocol 적용

chekInputError를 InputView에서 LadderError로 이동
LadderError에 에러체크를 위한 protocol 적용

june20516 committed:
LadderStep 중복 체크 개선

LadderStep.stepDecision 초기 랜덤값 가지도록 변경
LadderStep 내부에 decideStep method 생성하여 LadderGame에서 중복 확인 로직 이동
LadderGame의 property 접근범위 수정

june20516 committed:
PrintView 변수명 수정, handleError 함수 수정

PrintView.orderPlayerNames을 의미를 통일하기 위하여 printPlayerNames로 변경
main의 handleError 함수 내 인스턴스 선언 방식 변경
main함수의 값 전달 방식 변경

godrm commented on LadderGame/LadderGame/InputView.swift:
```diff
@@ -52,17 +52,5 @@ struct InputView {
-        }
-        return LadderGameError.none
-    }
+
```
Bool을 리턴하는 함수의 경우 is- 를 붙이는게 좀더 명확합니다. 참고하세요

godrm commented:
다음 단계를 하면서 지금까지 구현한 내용을 테스트할 수 있는지 확인해보고, 무엇을 테스트할 지 테스트함수를 만들어보세요
